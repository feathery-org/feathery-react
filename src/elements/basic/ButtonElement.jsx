import React, { useMemo, useState } from 'react';

import ReactButton from 'react-bootstrap/Button';
import Spinner from 'react-bootstrap/Spinner';
import TextNodes from '../components/TextNodes';

function adjustColor(color, amount) {
    return (
        '#' +
        color
            .replace(/^#/, '')
            .replace(/../g, (color) =>
                (
                    '0' +
                    Math.min(
                        255,
                        Math.max(0, parseInt(color, 16) + amount)
                    ).toString(16)
                ).substr(-2)
            )
    );
}

function applyButtonStyles(element, applyStyles) {
    applyStyles.addTargets('button', 'buttonActive', 'buttonHover', 'spinner');

    applyStyles.apply('button', 'background_color', (a) => ({
        backgroundColor: `#${a}`
    }));
    applyStyles.applyHeight('button');
    applyStyles.applyWidth('button');
    applyStyles.applyCorners('button');
    applyStyles.applyBorders('button');
    applyStyles.applyMargin('button');

    applyStyles.applyBorders('buttonHover', 'hover_');
    if (element.link !== 'none') {
        applyStyles.apply('buttonHover', 'background_color', (a) => {
            const color = `${adjustColor(a, -30)} !important`;
            return {
                backgroundColor: color,
                borderColor: color,
                transition: 'background 0.3s !important'
            };
        });
    }
    if (element.styles.hover_background_color) {
        applyStyles.apply('buttonHover', 'hover_background_color', (a) => ({
            backgroundColor: `#${a} !important`
        }));
    }

    applyStyles.applyBorders('buttonActive', 'selected_');
    if (element.styles.selected_background_color) {
        applyStyles.apply('buttonHover', 'selected_background_color', (a) => ({
            backgroundColor: `#${a} !important`
        }));
    }

    applyStyles.apply('spinner', 'show_spinner_on_submit', (a) => ({
        display: a ? 'default' : 'none'
    }));
    applyStyles.apply('spinner', ['height', 'height_unit'], (a, b) => {
        const thirdHeight = Math.round(a / 3);
        return {
            right: `-${a}${b}`,
            width: `${thirdHeight}${b}`,
            height: `${thirdHeight}${b}`
        };
    });

    return applyStyles;
}

function ButtonElement({
    element,
    applyStyles,
    values = null,
    handleRedirect = () => {},
    submit = () => {},
    onRepeatClick = () => {},
    setSubmitRef = () => {}
}) {
    const [showSpinner, setShowSpinner] = useState(false);

    async function buttonOnClick() {
        if (element.link === 'none') {
            return;
        }

        if (
            ['add_repeated_row', 'remove_repeated_row'].includes(element.link)
        ) {
            onRepeatClick();
            return;
        }

        if (element.link === 'submit') {
            setShowSpinner(true);
        }

        // Perform the submit callback
        // Note: We only need to set the spinner if the submit request failed
        // If the request succeeded then the button unmounts and calling setShowSpinner is an error
        try {
            const metadata = {
                elementType: 'button',
                elementIDs: [element.id],
                trigger: 'click'
            };
            let newStep;
            if (element.link === 'submit') {
                newStep = await submit(metadata, element.repeat || 0);
            } else {
                newStep = handleRedirect({ metadata });
            }

            // If the submit failed we want to throw here to turn off the spinner
            if (!newStep) setShowSpinner(false);
        } catch {
            setShowSpinner(false);
        }
    }
    if (element.link === 'submit') setSubmitRef(buttonOnClick);

    const styles = useMemo(() => applyButtonStyles(element, applyStyles), [
        applyStyles
    ]);
    return (
        <ReactButton
            id={element.id}
            key={element.id}
            style={{
                display: 'flex',
                justifyContent: 'center',
                alignItems: 'center',
                cursor: element.link === 'none' ? 'default' : 'pointer',
                boxShadow: 'none',
                maxWidth: '100%'
            }}
            css={{
                '&:disabled': {
                    cursor: 'default !important'
                },
                '&:active': styles.getTarget('buttonActive'),
                '&:hover:enabled': styles.getTarget('buttonHover'),
                ...styles.getTarget('button')
            }}
            disabled={element.link === 'none' || showSpinner}
            onClick={buttonOnClick}
        >
            <div style={{ display: 'flex', position: 'relative' }}>
                <TextNodes
                    element={element}
                    values={values}
                    applyStyles={applyStyles}
                    handleRedirect={handleRedirect}
                />
                {element.image_url && (
                    <img
                        src={element.image_url}
                        style={{
                            objectFit: 'contain',
                            width: '100%',
                            height: '100%'
                        }}
                    />
                )}
                {showSpinner && (
                    <Spinner
                        animation='border'
                        style={{
                            color: 'white',
                            position: 'absolute',
                            top: '50%',
                            bottom: '50%',
                            marginTop: 'auto',
                            marginBottom: 'auto',
                            border: '0.2em solid currentColor',
                            borderRightColor: 'transparent'
                        }}
                        css={{
                            ...styles.getTarget('spinner'),
                            '@-webkit-keyframes spinner-border': {
                                to: {
                                    WebkitTransform: 'rotate(360deg)',
                                    transform: 'rotate(360deg)'
                                }
                            },
                            '@keyframes spinner-border': {
                                to: {
                                    WebkitTransform: 'rotate(360deg)',
                                    transform: 'rotate(360deg)'
                                }
                            },
                            '&.spinner-border': {
                                borderRadius: '50%',
                                animation:
                                    '0.75s linear infinite spinner-border'
                            }
                        }}
                    />
                )}
            </div>
        </ReactButton>
    );
}

export default ButtonElement;
export { adjustColor };
