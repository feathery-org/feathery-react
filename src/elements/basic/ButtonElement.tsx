import React, { useMemo } from 'react';

import ReactButton from 'react-bootstrap/Button';
import TextNodes from '../components/TextNodes';
import { imgMaxSizeStyles } from '../styles';
import { adjustColor } from '../../utils/styles';
import useBorder from '../components/useBorder';
import { SUBMITTABLE_ACTIONS } from '../../utils/elementActions';

function applyButtonStyles(element: any, responsiveStyles: any) {
  responsiveStyles.addTargets(
    'button',
    'buttonActive',
    'buttonHover',
    'buttonDisabled',
    'loader',
    'img'
  );

  responsiveStyles.applyColor('button', 'background_color', 'backgroundColor');
  responsiveStyles.applyHeight('button');
  responsiveStyles.apply(
    'button',
    ['width', 'width_unit', 'content_responsive'],
    (a: any, b: any, c: boolean) => ({
      [c ? 'minWidth' : 'width']: `${a}${b}`
    })
  );
  responsiveStyles.applyCorners('button');
  responsiveStyles.applyFlexAndTextAlignments('button');
  responsiveStyles.apply(
    'button',
    [
      'uploader_padding_top',
      'uploader_padding_right',
      'uploader_padding_bottom',
      'uploader_padding_left'
    ],
    // @ts-expect-error TS(7006): Parameter 'a' implicitly has an 'any' type.
    (a, b, c, d) => ({
      padding: `${a}px ${b}px ${c}px ${d}px`
    })
  );
  responsiveStyles.applyWidth('img', 'image_');
  responsiveStyles.applyMargin('img', 'image_');

  if (element.styles.hover_background_color) {
    responsiveStyles.applyColor(
      'buttonHover',
      `hover_background_color`,
      'backgroundColor',
      true
    );
  } else {
    // default hover effect
    responsiveStyles.apply('buttonHover', 'background_color', (a: any) => {
      const newColor = `${adjustColor(a, -45)} !important`;
      return { backgroundColor: newColor };
    });
  }
  responsiveStyles.applySpanSelectorStyles('buttonHover', 'hover_');
  responsiveStyles.apply('buttonHover', 'hover_image_color', (a: string) => {
    if (!a) return {};
    const level = a === 'black' ? 0 : 100;
    return {
      img: {
        webkitFilter: `brightness(${level}%)`,
        filter: `brightness(${level}%)`
      }
    };
  });

  responsiveStyles.applyColor(
    'buttonActive',
    `selected_background_color`,
    'backgroundColor',
    true
  );
  responsiveStyles.applySpanSelectorStyles('buttonActive', 'selected_');
  responsiveStyles.apply(
    'buttonActive',
    'selected_image_color',
    (a: string) => {
      if (!a) return {};
      const level = a === 'black' ? 0 : 100;
      return {
        img: {
          webkitFilter: `brightness(${level}%)`,
          filter: `brightness(${level}%)`
        }
      };
    }
  );

  responsiveStyles.apply('buttonDisabled', 'background_color', (a: any) => {
    const color = `${adjustColor(a, 45)} !important`;
    return {
      backgroundColor: color,
      borderColor: color
    };
  });
  responsiveStyles.apply(
    'buttonDisabled',
    'disabled_image_color',
    (a: string) => {
      if (!a) return {};
      const level = a === 'black' ? 0 : 100;
      return {
        img: {
          webkitFilter: `brightness(${level}%)`,
          filter: `brightness(${level}%)`
        }
      };
    }
  );
  responsiveStyles.applySpanSelectorStyles('buttonDisabled', 'disabled_');
  responsiveStyles.applyColor(
    'buttonDisabled',
    `disabled_background_color`,
    'backgroundColor',
    true
  );

  responsiveStyles.apply(
    'loader',
    ['height', 'height_unit'],
    (a: any, b: any) => {
      const thirdHeight = Math.round(a / 3);
      const dimension = `${thirdHeight}${b}`;
      return { width: dimension, height: dimension };
    }
  );

  return responsiveStyles;
}

function ButtonElement({
  element,
  responsiveStyles,
  loader = null,
  editMode,
  focused = false,
  disabled = false,
  active = false,
  textCallbacks = {},
  onClick = () => {},
  elementProps = {},
  inlineError,
  children
}: any) {
  const styles = useMemo(
    () => applyButtonStyles(element, responsiveStyles),
    [responsiveStyles]
  );
  const { borderStyles, customBorder } = useBorder({
    element,
    error: inlineError,
    defaultHover: true
  });

  const activeStyles = editMode
    ? styles.getTarget('button')
    : {
        ...styles.getTarget('buttonActive'),
        ...borderStyles.active
      };

  const actions = element.properties.actions;
  // type=submit is important for HTML5 type validation messages
  const type = actions.some(
    (action: any) => SUBMITTABLE_ACTIONS.includes(action.type) && action.submit
  )
    ? 'submit'
    : 'button';
  return (
    <ReactButton
      id={element.id}
      key={element.id}
      active={active}
      type={type}
      style={{
        display: 'flex',
        cursor: editMode || actions.length === 0 ? 'default' : 'pointer',
        boxShadow: 'none',
        maxWidth: '100%',
        position: 'relative'
      }}
      css={{
        justifyContent: 'center',
        alignItems: 'center',
        border: 'none',
        transition: '0.3s ease all !important',
        '&:disabled': {
          cursor: 'default !important',
          ...styles.getTarget('buttonDisabled'),
          ...borderStyles.disabled
        },
        '&:hover:enabled': editMode
          ? styles.getTarget('button')
          : {
              ...styles.getTarget('buttonHover'),
              ...borderStyles.hover
            },
        '&.active:enabled': activeStyles,
        '&:focus:enabled': activeStyles,
        '&&': styles.getTarget('button')
      }}
      disabled={actions.length === 0 || loader || disabled}
      onClick={onClick}
      {...elementProps}
    >
      {customBorder}
      {children}
      {loader ? (
        <div css={styles.getTarget('loader')}>{loader}</div>
      ) : (
        <>
          {element.properties.image && (
            <img
              src={element.properties.image}
              style={{
                ...imgMaxSizeStyles,
                ...responsiveStyles.getTargets('img')
              }}
            />
          )}
          {element.properties.text && (
            <TextNodes
              element={element}
              responsiveStyles={responsiveStyles}
              editMode={editMode}
              focused={focused}
              textCallbacks={textCallbacks}
            />
          )}
        </>
      )}
    </ReactButton>
  );
}

export default ButtonElement;
