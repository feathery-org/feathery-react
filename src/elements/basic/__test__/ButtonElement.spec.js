import React from 'react';
import { create, act } from 'react-test-renderer';
import { adjustColor } from '../ButtonElement';
import Elements from '../..';

describe('ButtonElement', () => {
    describe('adjustColor', () => {
        it('adjusts color up', () => {
            // Arrange
            const color = '#000000';
            const amount = 30;
            const expected = '#1e1e1e';

            // Act
            const actual = adjustColor(color, amount);

            // Assert
            expect(actual).toEqual(expected);
        });

        it('adjusts color down', () => {
            // Arrange
            const color = '#ffffff';
            const amount = -30;
            const expected = '#e1e1e1';

            // Act
            const actual = adjustColor(color, amount);

            // Assert
            expect(actual).toEqual(expected);
        });
    });

    it('renders an empty ButtonElement element', async () => {
        // Arrange
        const props = {
            element: {
                text_formatted: [{ insert: '\n' }],
                column_index: 0,
                row_index: 0,
                column_index_end: 0,
                row_index_end: 0,
                styles: {
                    border_top_color: '2954af',
                    border_right_color: '2954af',
                    border_bottom_color: '2954af',
                    border_left_color: '2954af',
                    background_color: '2954af',
                    layout: 'center'
                },
                mobile_styles: {}
            },
            values: {},
            submit: () => {},
            setSubmitRef: () => {},
            onRepeatClick: () => {}
        };

        // Act
        let button;
        act(() => {
            button = create(<Elements.ButtonElement {...props} />);
        });
        const tree = button.toJSON();

        // Assert
        expect(tree).toMatchSnapshot();
    });

    it('renders a button with plain text', async () => {
        // Arrange
        const props = {
            element: {
                text_formatted: [{ insert: 'Hello World!' }],
                column_index: 0,
                row_index: 0,
                column_index_end: 0,
                row_index_end: 0,
                styles: {
                    background_color: '2954af',
                    layout: 'center'
                },
                mobile_styles: {}
            },
            values: {},
            submit: () => {},
            setSubmitRef: () => {},
            onRepeatClick: () => {}
        };

        // Act
        let button;
        act(() => {
            button = create(<Elements.ButtonElement {...props} />);
        });
        const tree = button.toJSON();

        // Assert
        expect(tree).toMatchSnapshot();
    });

    it('renders a button with custom styles', async () => {
        // Arrange
        const props = {
            element: {
                text_formatted: [
                    {
                        insert: 'Hello World!',
                        attributes: {
                            weight: 700,
                            color: '#000000FF',
                            size: '48px',
                            font: 'sans-serif',
                            italic: true,
                            strike: false,
                            underline: false
                        }
                    }
                ],
                column_index: 0,
                row_index: 0,
                column_index_end: 0,
                row_index_end: 0,
                styles: {
                    background_color: '2954af',
                    layout: 'center'
                },
                mobile_styles: {}
            },
            values: {},
            submit: () => {},
            setSubmitRef: () => {},
            onRepeatClick: () => {}
        };

        // Act
        let button;
        act(() => {
            button = create(<Elements.ButtonElement {...props} />);
        });
        const tree = button.toJSON();

        // Assert
        expect(tree).toMatchSnapshot();
    });
});
