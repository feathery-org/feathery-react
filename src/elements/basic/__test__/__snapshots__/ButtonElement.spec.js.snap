// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonElement renders a button with custom styles 1`] = `
<button
  className="css-100jsw4-ButtonElement btn btn-primary"
  disabled={false}
  onClick={[Function]}
  style={
    Object {
      "alignItems": "center",
      "boxShadow": "none",
      "cursor": "pointer",
      "display": "flex",
      "justifyContent": "center",
      "maxWidth": "100%",
    }
  }
  type="button"
>
  <div
    style={
      Object {
        "display": "flex",
        "position": "relative",
      }
    }
  >
    <span
      className="css-1kxrhf3"
      onClick={[Function]}
    >
      Hello World!
    </span>
  </div>
</button>
`;

exports[`ButtonElement renders a button with plain text 1`] = `
<button
  className="css-100jsw4-ButtonElement btn btn-primary"
  disabled={false}
  onClick={[Function]}
  style={
    Object {
      "alignItems": "center",
      "boxShadow": "none",
      "cursor": "pointer",
      "display": "flex",
      "justifyContent": "center",
      "maxWidth": "100%",
    }
  }
  type="button"
>
  <div
    style={
      Object {
        "display": "flex",
        "position": "relative",
      }
    }
  >
    <span
      className="css-1kxrhf3"
      onClick={[Function]}
    >
      Hello World!
    </span>
  </div>
</button>
`;

exports[`ButtonElement renders an empty ButtonElement element 1`] = `
<button
  className="css-89fnp8-ButtonElement btn btn-primary"
  disabled={false}
  onClick={[Function]}
  style={
    Object {
      "alignItems": "center",
      "boxShadow": "none",
      "cursor": "pointer",
      "display": "flex",
      "justifyContent": "center",
      "maxWidth": "100%",
    }
  }
  type="button"
>
  <div
    style={
      Object {
        "display": "flex",
        "position": "relative",
      }
    }
  >
    <span
      className="css-1kxrhf3"
      onClick={[Function]}
    >
      

    </span>
  </div>
</button>
`;
